### ISSUE TITLE
Evaluators versioned by hardfork

### ISSUE THREAD

**Issue by youkaicountry** [steemit/steem#2544]
Rather than having `if` hardfork clauses within each evaluator, implement separate evaluators versioned by hardfork.

**Comment by TimCliff** [comment]
There has been some discussion among the witnesses about breaking away from using sequential hardfork numbers to distinguish future code changes. The reason being that we will likely reach a point where multiple hardfork changes are being worked on in parallel, without necessarily knowing what order they will get completed in.

A good example of this is SMTs. Currently all of the SMT code is under hardfork 21. If we assume that “Velocity” ends up being HF20, but we want to implement some hardfork change after 20 but before SMTs, it will become difficult and confusing because now we have all of the “hardfork 21” (SMTs) code that is no longer going to be “21” to deal with.

A better approach seems to be to use a feature based naming convention (such as “HF_SMT”) until all of the development is done, and then assign the HF number once everything is ready and it is chosen/decided to be the next hardfork.

Is this out of scope for this issue, or something that could be addressed at the same time?

**Comment by mvandeberg** [comment]
Do you mean like this? https://github.com/steemit/steem/blob/master/libraries/protocol/hardfork.d/0_21.hf#L3

**Comment by TimCliff** [comment]
Yes, along those lines. It would probably need some additional refactoring though. For example, that code is currently defined in the `0_21.hf` file. If we weren't ready to assign the SMT HF a number yet, then `0_21.hf` would not be the appropriate place for that code.

Also, the 'hardfork 20' code is probably a little easier to talk about as far as what not to do anymore. Even though the hardfork is not completed yet, all of the specific changes for Velocity are defined here:
https://github.com/steemit/steem/blob/master/libraries/protocol/hardfork.d/0_20.hf
![image](https://user-images.githubusercontent.com/20735105/41419457-b4e92634-6fb7-11e8-8fa4-177f6fdae55b.png)

Throughout the code, there are checks like this:
https://github.com/steemit/steem/blob/62b48877d9f731c3fe00ef818e3324a0a3de3e63/libraries/chain/database.cpp#L3597

If we hypothetically wanted to do a different HF first and make all of the Velocity changes "hardfork 21", then there would be a lot of things to update in the code.
